File: geomesa-hbase/geomesa-hbase-datastore/src/main/java/org/locationtech/geomesa/hbase/driver/KryoLazyDensityDriver.java
/***********************************************************************
* Copyright (c) 2013-2016 Commonwealth Computer Research, Inc.
* All rights reserved. This program and the accompanying materials
* are made available under the terms of the Apache License, Version 2.0
* which accompanies this distribution and is available at
* http://www.opensource.org/licenses/apache2.0.php.
*************************************************************************/

package org.locationtech.geomesa.hbase.driver;

import com.google.protobuf.ByteString;
import com.google.protobuf.RpcCallback;
import com.google.protobuf.RpcController;
import org.apache.hadoop.hbase.client.Table;
import org.apache.hadoop.hbase.client.coprocessor.Batch.*;
import org.apache.hadoop.hbase.ipc.BlockingRpcCallback;
import org.locationtech.geomesa.hbase.proto.KryoLazyDensityProto.*;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class KryoLazyDensityDriver {

    /**
     * An RpcController implementation for use here in this endpoint.
     */
    static class KryoLazyDensityRpcController implements RpcController {
        private String errorText;
        private boolean cancelled = false;
        private boolean failed = false;

        @Override
        public String errorText() {
            return this.errorText;
        }

        @Override
        public boolean failed() {
            return this.failed;
        }

        @Override
        public boolean isCanceled() {
            return this.cancelled;
        }

        @Override
        public void notifyOnCancel(RpcCallback<Object> arg0) {
            throw new UnsupportedOperationException();
        }

        @Override
        public void reset() {
            this.errorText = null;
            this.cancelled = false;
            this.failed = false;
        }

        @Override
        public void setFailed(String errorText) {
            this.failed = true;
            this.errorText = errorText;
        }

        @Override
        public void startCancel() {
            this.cancelled = true;
        }

    }

    /**
     * It gives the combined result of pairs received from different region servers value of a column for a given column family for the
     * given range. In case qualifier is null, a min of all values for the given
     * family is returned.
     *
     * @param table
     * @return HashMap result;
     * @throws Throwable
     */
    public List<ByteString> kryoLazyDensityFilter(final Table table, final byte[] filter) throws Throwable {

        final DensityRequest requestArg = DensityRequest.newBuilder().setByteFilter(ByteString.copyFrom(filter)).build();
        class KryoLazyDensityFilterCallBack implements Callback<ByteString> {

            private List<ByteString> finalResult = new ArrayList<>();

            List<ByteString> getResult() {
                List<ByteString> list = new ArrayList<ByteString>();
                for (ByteString s : finalResult) {
                    list.add(s);
                }
                return list;
            }

            @Override
            public synchronized void update(byte[] region, byte[] row, ByteString result) {
                finalResult.add(result);
            }
        }

        KryoLazyDensityFilterCallBack kryoLazyDensityFilterCallBack = new KryoLazyDensityFilterCallBack();
        table.coprocessorService(KryoLazyDensityService.class, null, null,
                new Call<KryoLazyDensityService, ByteString>() {
                    @Override
                    public ByteString call(KryoLazyDensityService instance) throws IOException {
                        RpcController controller = new KryoLazyDensityRpcController();
                        BlockingRpcCallback<DensityResponse> rpcCallback =
                                new BlockingRpcCallback<>();
                        instance.getDensity(controller, requestArg, rpcCallback);
                        DensityResponse response = rpcCallback.get();
                        if (controller.failed()) {
                            throw new IOException(controller.errorText());
                        }

                        return response.getSf();
                    }
                }, kryoLazyDensityFilterCallBack);

        return kryoLazyDensityFilterCallBack.getResult();
    }
}

File: geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/driver/KryoLazyDensityDriver.scala
/***********************************************************************
 * Copyright (c) 2013-2017 Commonwealth Computer Research, Inc.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Apache License, Version 2.0
 * which accompanies this distribution and is available at
 * http://www.opensource.org/licenses/apache2.0.php.
 ***********************************************************************/

package org.locationtech.geomesa.hbase.driver

import java.io.IOException
import java.util.concurrent.ConcurrentLinkedQueue
import java.util.Queue

import com.google.protobuf.{ByteString, RpcCallback, RpcController}
import org.apache.hadoop.hbase.client.Table
import org.apache.hadoop.hbase.client.coprocessor.Batch._
import org.apache.hadoop.hbase.ipc.BlockingRpcCallback
import org.locationtech.geomesa.hbase.proto.KryoLazyDensityProto._

import scala.collection.JavaConversions._

/**
  * An RpcController implementation for use here in this endpoint.
  */
class KryoLazyDensityRpcController extends RpcController {

  private[driver] var errorText: String = _

  private var cancelled: Boolean = false

  private[driver] var failed: Boolean = false

  override def isCanceled: Boolean = this.cancelled

  override def reset(): Unit = {
    this.errorText = null
    this.cancelled = false
    this.failed = false
  }

  override def setFailed(errorText: String): Unit = {
    this.failed = true
    this.errorText = errorText
  }

  override def startCancel(): Unit = {
    this.cancelled = true
  }

  override def notifyOnCancel(rpcCallback: RpcCallback[AnyRef]): Unit = {
    throw new UnsupportedOperationException()
  }
}

class KryoLazyDensityDriver {

  /**
    * It gives the combined result of pairs received from different region servers value of a column for a given column family for the
    * given range. In case qualifier is null, a min of all values for the given
    * family is returned.
    *
    * @param table
    * @return HashMap result;
    * @throws Throwable
    */
  def kryoLazyDensityFilter(table: Table, options: Array[Byte]): List[ByteString] = {
    val requestArg: DensityRequest = DensityRequest.newBuilder().setOptions(ByteString.copyFrom(options)).build()

    class KryoLazyDensityFilterCallBack extends Callback[ByteString] {

      private val finalResult: Queue[ByteString] = new ConcurrentLinkedQueue[ByteString]()

      def getResult: List[ByteString] = {
        var list: List[ByteString] = List[ByteString]()
        for (s <- finalResult) {
          list ::= s
        }
        list
      }

      override def update(region: Array[Byte], row: Array[Byte], result: ByteString): Unit = {
        finalResult.offer(result)
      }
    }

    val kryoLazyDensityFilterCallBack: KryoLazyDensityFilterCallBack = new KryoLazyDensityFilterCallBack()
    table.coprocessorService(classOf[KryoLazyDensityService], null, null, new Call[KryoLazyDensityService, ByteString]() {
      override def call(instance: KryoLazyDensityService): ByteString = {
        val controller: RpcController = new KryoLazyDensityRpcController()
        val rpcCallback: BlockingRpcCallback[DensityResponse] =
          new BlockingRpcCallback[DensityResponse]()
        instance.getDensity(controller, requestArg, rpcCallback)
        val response: DensityResponse = rpcCallback.get
        if (controller.failed()) {
          throw new IOException(controller.errorText())
        }
        response.getSf
      }
    },
      kryoLazyDensityFilterCallBack
    )
    kryoLazyDensityFilterCallBack.getResult
  }

}
